k2/csrc/fsa_equivalent.cc

```diff
@@ -0,0 +1,82 @@
+// k2/csrc/fsa_equivalent.cc
+
+// Copyright (c)  2020 Haowen Qiu
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/fsa_equivalent.h"
+
+#include <algorithm>
+#include <random>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "k2/csrc/fsa.h"
+// #include "k2/csrc/fsa_algo.h"
+#include "k2/csrc/properties.h"
+
+namespace k2 {
+
+bool RandomPath(const Fsa &a, Fsa *b, std::vector<int32_t> *state_map) {
+  if (IsEmpty(a) || b == nullptr) return false;
+  if (!IsConnected(a)) {
+    Fsa connected_fsa;
+    // TODO(haowen): uncomment this
+    // Connect(a, &connected_fsa);
+    return RandomPath(connected_fsa, b, state_map);
+  }
+
+  int32_t num_states = a.NumStates();
+  std::vector<int32_t> state_map_b2a;
+  std::vector<int32_t> state_map_a2b(num_states, -1);
+  // `visited_arcs[i]` stores `arcs` leaving from state `i`
+  std::vector<std::unordered_set<Arc, ArcHash>> visited_arcs;
+
+  std::random_device rd;
+  std::mt19937 generator(rd());
+  std::uniform_int_distribution<int32_t> distribution(0);
+
+  int32_t num_visited_arcs = 0;
+  int32_t num_visited_state = 0;
+  int32_t state = 0;
+  int32_t final_state = num_states - 1;
+  while (true) {
+    if (state_map_a2b[state] == -1) {
+      state_map_a2b[state] = num_visited_state;
+      state_map_b2a.push_back(state);
+      visited_arcs.push_back(std::unordered_set<Arc, ArcHash>());
+      ++num_visited_state;
+    }
+    if (state == final_state) break;
```

Comment:
It **might** not terminate and be an infinite loop here if there is a cycle in
the graph.

Reply:
If the input fsa is connected, probabilistically the algorithm will always terminate in finite time.